src/contextmenu.js

```diff
@@ -0,0 +1,472 @@
+ * multiple blocks.
+ */
+const registerDuplicate = function() {
+  const duplicateOption = {
+    displayText: function() {
+      const selectedBlocksLength = blockSelection.size;
+      if (selectedBlocksLength <= 1) {
+        return Blockly.Msg['DUPLICATE_BLOCK'];
+      } else {
+        return Blockly.Msg['DUPLICATE_BLOCK'] + ' (' +
+        selectedBlocksLength + ')';
+      }
+    },
+    preconditionFn: function(scope) {
+      const block = scope.block;
+      if (!block.isInFlyout && block.isDeletable() && block.isMovable()) {
+        if (block.isDuplicatable()) {
+          return 'enabled';
+        }
+        return 'disabled';
+      }
+      return 'hidden';
+    },
+    callback: function(scope) {
+      const duplicatedBlocks = [];
+      const apply = function(block) {
+        if (block && duplicateOption.preconditionFn({
+          block: block,
+        }) === 'enabled' && !hasSelectedParent(block)) {
+          duplicatedBlocks.push(Blockly.clipboard.duplicate(block));
+        }
+      };
+      Blockly.Events.setGroup(true);
+      if (scope.block && blockSelection.size === 0) {
+        apply(scope.block);
+      }
+      blockSelection.forEach(function(id) {
+        const block = Blockly.getMainWorkspace().getBlockById(id);
+        if (block) {
+          apply(block);
+          block.pathObject.updateSelected(false);
+        }
+      });
+      Blockly.Events.setGroup(false);
+      blockSelection.clear();
+      duplicatedBlocks.forEach(function(block) {
+        if (block.id) {
+          blockSelection.add(block.id);
+          block.pathObject.updateSelected(true);
+        }
+      });
+    },
+    scopeType: Blockly.ContextMenuRegistry.ScopeType.BLOCK,
+    id: 'blockDuplicate',
+    weight: 1,
+  };
+  Blockly.ContextMenuRegistry.registry.register(duplicateOption);
+};
+
+/**
+ * Modification for context menu 'Comment' to be available for multiple blocks.
+ */
+const registerComment = function() {
+  const commentOption = {
+    displayText: function(scope) {
+      const selectedBlocksLength = blockSelection.size;
+      if (scope.block.getCommentIcon()) {
+        if (selectedBlocksLength <= 1) {
+          return Blockly.Msg['REMOVE_COMMENT'];
+        } else {
+          return Blockly.Msg['REMOVE_COMMENT'] + ' (' +
+          selectedBlocksLength + ')';
+        }
+      }
+      if (selectedBlocksLength <= 1) {
+        return Blockly.Msg['ADD_COMMENT'];
+      } else {
+        return Blockly.Msg['ADD_COMMENT'] + ' (' +
+        selectedBlocksLength + ')';
+      }
+    },
+    preconditionFn: function(scope) {
+      const block = scope.block;
+      if (!Blockly.utils.userAgent.IE && !block.isInFlyout &&
+        block.workspace.options.comments && !block.isCollapsed() &&
+        block.isEditable()) {
+        return 'enabled';
+      }
+      return 'hidden';
+    },
+    callback: function(scope) {
+      const hasCommentIcon = scope.block.getCommentIcon();
```

Comment:
I like determining whether comments get added or removed based on the block that was right-clicked! Very clever =)